katdal/applycal.py

```diff
@@ -291,6 +310,9 @@ def calc_correction_per_input(cache, name, inp, product_type):
                                                          data_freqs, cal_freqs)
         elif product_type == 'G':
             correction_sensor = calc_gain_correction(product_sensor, index)
+        elif product_type in ('GPHASE', 'GAMP_PHASE'):
+            targets = cache.get('Observation/target_index')
+            correction_sensor = calc_gain_correction(product_sensor, index, targets)
```

Comment:
I agree that it is a bit clunky. In the meantime, discussions with the pipeline team revealed that, in fact, *all* cal products are target-aware, so this key might not be needed after all...
The truth is even messier and would need another PR :sad:

What we have here could still tick the L2 box, so I'm proposing no further changes within this PR.

Reply:
If we want to drive these decisions from telstate then we need to get it figured out ASAP. If we leave it until L2 is "done" and we start making using it, it'll be difficult to change because we'll have files in the archive without the necessary telstate keys.

Reply:
As it stands with this PR, all images of target fields will be fine (the main cause), even without the keys. If you want to image your gain calibrators, that will also be fine. Only images of the bandpass calibrators will be slightly off. The real kicker would be once flux calibration gets thrown in the mix.

I agree with your general sentiment though.